infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/Orchestrator.java

```diff
@@ -34,37 +41,73 @@
  * that is specified by Workflow.getActions() to achieve a bigger goal. For example, growing the
  * cluster. Initiated through RPC, the orchestrator will create a workflow instance and attempt
  * to execute all its actions.
- *
+ * <p>
  * Created by Maithem on 10/25/17.
  */
 
 @Slf4j
 public class Orchestrator {
 
+    /**
+     * The number of times to retry an action before failing the workflow.
+     */
+    private static final int ACTION_RETRY = 3;
+
     final ServerContext serverContext;
 
     final Callable<CorfuRuntime> getRuntime;
 
     final BiMap<UUID, String> activeWorkflows = Maps.synchronizedBiMap(HashBiMap.create());
 
+    final ExecutorService executor;
+
     public Orchestrator(@Nonnull Callable<CorfuRuntime> runtime,
                         @Nonnull ServerContext serverContext) {
         this.serverContext = serverContext;
         this.getRuntime = runtime;
+
+        executor = Executors.newFixedThreadPool(Runtime.getRuntime()
+                .availableProcessors(), new ThreadFactory() {
+
+            final AtomicInteger threadNumber = new AtomicInteger(0);
+
+            @Override
+            public Thread newThread(Runnable r) {
+                Thread thread = new Thread(r);
+                thread.setDaemon(true);
+                String threadName = serverContext.getThreadPrefix() + "orchestrator-"
```

Comment:
Weren't you going to use the workflow id (UuidUtils.toBase64 String...)

Reply:
This is a fixed size pool, a thread is not spawned per workflow.